<h1> Step 6: Alpha-Beta Pruning Algorithm </h1>

In [ ]:
from src.games import TicTacToe

from src.alphabeta import (
    alpha_beta_value,
    alpha_beta_decision,
    alpha_beta_value_with_stats,
    alpha_beta_decision_ordered,
    alpha_beta_value_ordered_with_stats,
)
from src.metrics import (
    SearchStats
)
from src.minimax import (
    minimax_value,
)

Alpha-Beta says: </br>
If a part of the tree has no other impact on the final decision, don't examine it.

That is:</br>
Pruning = removing unnecessary branches

**Alpha (α):** </br>
The best guaranteed value for MAX so far.</br>
That is: The minimum value MAX can take is this.

**Beta (β):**</br>
The best guaranteed value for MIN.</br>
That is: The maximum value MIN allows is this.

Alpha-Beta = Minimax + Pruning

In [2]:
board1 = (
    "X", "X", " ",
    "O", "O", " ",
    " ", " ", " ",
)

board2 = (
    "X", " ", " ",
    "O", "O", " ",
    "X", " ", " ",
)

board3 = (
    "O", "O", " ",
    "X", "X", " ",
    " ", " ", " ",
)

In [3]:
game = TicTacToe()

game

In [4]:
state = (
    board3,
    "X",
)

alpha beta value test on a winning State → value 10000

In [5]:
value = alpha_beta_value(
    game,
    state,
    alpha=-float("inf"),
    beta=float("inf"),
)

value

10000

alpha beta decision test on a specific State → Valid Action

In [6]:
action = alpha_beta_decision(
    game,
    state,
)

action

2

alpha beta decision test on initial State (Game start test) → Valid Action

In [7]:
state = game.initial_state()

alpha_beta_decision(
    game,
    state,
)

0

Why does the order of actions affect Alpha-Beta? </br>
Because Alpha-Beta will prune more if it sees good moves earlier.

<h3> Step 6.1: Measuring Alpha-Beta Pruning Effect </h3>

But the main purpose of Alpha-Beta: </br>
Reducing the number of checked nodes without changing is the answer.

In [8]:
game = TicTacToe()

state = game.initial_state()

In [9]:
stats = SearchStats()

value = alpha_beta_value_with_stats(
    game,
    state,
    alpha=-float("inf"),
    beta=float("inf"),
    stats=stats,
)

stats.summary()

{'Nodes Visited': 18297, 'Terminal States': 7330, 'Max Depth': 0}

<h3> Step 6.2: Comparing with complete min-max </h3>

**complete Minimax** </br>
Nodes Visited:     549945</br>
Terminal States:   255168</br>
Max Depth:         9


**Alpha-Beta**</br>
Nodes Visited:     18297</br>
Terminal States:   7330</br>
Max Depth:         0

*About 96.6% reduction in search space*

In [10]:
minimax_value_result = minimax_value(
    game,
    state,
)

alpha_beta_value_result = alpha_beta_value(
    game,
    state,
    alpha=-float("inf"),
    beta=float("inf"),
)


print(minimax_value_result)
print(alpha_beta_value_result)

0
0


In Tic-Tac-Toe from the starting position: </br>

If both players play optimally:</br>
X cannot force a win.</br>
O cannot force a win either.

The optimal outcome of the game is: Draw.

And in our utility:</br>
Draw = 0

In [11]:
state = (
    board3,
    "X",
)

minimax_value_result = minimax_value(
    game,
    state,
)

alpha_beta_value_result = alpha_beta_value(
    game,
    state,
    alpha=-float("inf"),
    beta=float("inf"),
)

print(minimax_value_result)
print(alpha_beta_value_result)

10000
10000


We checked:

*Minimax(s)=AlphaBeta(s)*

That is:</br>
Alpha-Beta only reduces Search, not the result.

<h3> Step 6.3: Move Ordering and Best/Worst Case Alpha-Beta </h3>

Move Ordering is a concept within Alpha-Beta. </br>
That is, it is an optimization, not a new algorithm.

Goal:</br>
Alpha-Beta does the most pruning when:</br>
It checks good moves early.

**First Heuristic for Move Ordering**

For Tic-Tac-Toe:</br>

Priority:</br>
Immediate Winning Move</br>
Defensive Move</br>
Center</br>
Corners</br>
Rest

In [12]:
game = TicTacToe()

state = game.initial_state()

action = alpha_beta_decision_ordered(
    game,
    state,
)

action

4

In [13]:
state = (
    board3,
    "X",
)

alpha_beta_decision_ordered(
    game,
    state,
)

2

<h3> Step 6.4: Move Ordering + Stats </h3>

In [14]:
game = TicTacToe()

state = game.initial_state()

stats = SearchStats()

value = alpha_beta_value_ordered_with_stats(
    game,
    state,
    alpha=-float("inf"),
    beta=float("inf"),
    stats=stats,
)
print(value)

stats.summary()

0


{'Nodes Visited': 7275, 'Terminal States': 2893, 'Max Depth': 9}

| Algorithm             | Nodes Visited | Reduction compared to Minimax | Reduction compared to Alpha-Beta |
| --------------------- | ------------: | -------------------: | -------------------: |
| Minimax               |        549945 |                    — |                    — |
| Alpha-Beta            |         18297 |     About 96.6% less |                    — |
| Alpha-Beta + Ordering |          7275 |     About 98.7٪ less |       About 60٪ less |
